connector_salesforce/unit/exporter_synchronizer.py

```diff
@@ -0,0 +1,354 @@
+
+        :return: an record of Odoo model base on
+                 `_model_name` key
+        :rtype: :py:class:`openerp.osv.orm.BrowseRecord`
+        """
+        assert self.binding_id
+        return self.session.browse(
+            self._model_name,
+            self.binding_id
+        )
+
+    def _map_data_for_upsert(self, mapper, **kwargs):
+        """ Call the convert function of the Mapper
+
+        in order to get converted record by using
+        mapper.data or mapper.data_for_create
+
+        :return: mapped dic of data to be used by
+                 :py:meth:``models.Model.create``
+        :rtype: dict
+        """
+        data = mapper.values(**kwargs)
+        self._validate_data(data)
+        return data
+
+    def _upsert(self, salesforce_id, data):
+        """Implementation of upserting a record on
+        Salesforce.
+
+        upsert means update or create
+
+        :param salesforce_id: the id of Odoo binding to export
+        :type salesforce_id: int or long
+
+        :param data: a dict with key and values to be updated created
+
+        :return: the Salesforce uuid of created or updated record
+        """
+        sf_id = self.backend_adapter.upsert(salesforce_id, data)
+        return sf_id
+
+    def _validate_data(self, data):
+        """ Check if the values to export are correct
+
+        :param data: data to validate
+        Proactively check before the ``_upsert``
+        if some fields are missing or invalid.
+
+        Raise `InvalidDataError`
+        """
+        # we may want to return an NotImplementedError
+        return
+
+    def _export(self, binding_id, salesforce_id):
+        """Export a binding record on Salesforce using REST API
+        :param binding_id: the current binding id in Odoo
+        :type binding_id: int or long
+
+        :param salesforce_id: Salesforce uuid if available else None/False
+        :type salesforce_id: str on None
+
+        """
+        record_mapper = self.mapper.map_record(self.binding_record)
+        # optimisation trick to avoid lookup binding
+        data = self._map_data_for_upsert(record_mapper,
+                                         binding_id=binding_id,
+                                         backend_record=self.backend_record)
+        sf_id = self._upsert(salesforce_id, data)
+        self.binder.bind(sf_id, binding_id)
+
+    def run(self, binding_id, force_deactivate=False):
+        """Try to export or deactivate a record on Salesforce using REST API
+        call required hooks and bind the record
+
+        :param binding_id: the current binding id in Odoo
+        :type binding_id: int or long
+
+        :param force_deactivate: If set to True it will force deactivate
+                                 without calling _to_deactivate
+                                 mostly use to save some REST calls
+
+        """
+        self.binding_id = binding_id
+        self.binding_record = self._get_record()
```

Comment:
I told the binding, but it could be entire entities like the partner + its binding. In that case, there is nothing to export and it just just return without error.